### PR TITLE
Add quality check of calibration results

### PR DESCRIPTION
**Pull request by daniel-mills-cqc**
This adds a check on the quality of the calibration results. In particular it checks if a large fraction of the noisy expectation values of the calibration circuits are close to that of the original circuit. The training procedure is more accurate when a large fraction of the calibration circuits have expectation values close to that of the original.

### CODE REVIEW

File: qermit/clifford_noise_characterisation/ccl.py [hunk 1]
```diff
@@ -497,20 +497,27 @@ def gen_CDR_MitEx(
...
             kwargs.get("model", _PolyCDRCorrect(1)),
             kwargs.get("tolerance", 0.01),
         )
     )
 
-    _experiment_taskgraph = TaskGraph().from_TaskGraph(_experiment_mitex)
-    _experiment_taskgraph.parallel(_states_sim_taskgraph)
+    _post_task_graph = TaskGraph(_label="QualityCheckCorrect")
+    _post_task_graph.parallel(_post_calibrate_task_graph)
+    _post_task_graph.prepend(cdr_quality_check_task_gen(distance_tolerance=kwargs.get("distance_tolerance", 0.1)))
```
Comment (sjdilkes) [inline qermit/clifford_noise_characterisation/ccl.py:515]: please add this as a "key" parameter in documentation

File: qermit/clifford_noise_characterisation/cdr_post.py [hunk 2]
```diff
@@ -67,6 +71,76 @@ def correct(self, noisy_expectation: float) -> float:
...
+    :param distance_tolerance: The absolute tolerance on the distance between
+    expectation values of the calibration and original circuit.
+    :type distance_tolerance: float
+    """
+
+    def cdr_quality_check_task(
+        obj,
+        noisy_expectation: List[QubitPauliOperator],
+        state_circuit_exp: List[List[Tuple[QubitPauliOperator, QubitPauliOperator]]],
+    ):
```
Comment (sjdilkes) [inline qermit/clifford_noise_characterisation/cdr_post.py:89]: -> Tuple[whatever typing is]

File: qermit/clifford_noise_characterisation/cdr_post.py [hunk 3]
```diff
@@ -67,6 +71,76 @@ def correct(self, noisy_expectation: float) -> float:
...
+                # Check the closeness of the expectation value of the training
+                # circuit and that of the original circuit.
+                if not math.isclose(
+                    noisy_coefficient, original_coefficient, abs_tol=distance_tolerance
+                ):
+                    is_far_count += 1
+
+            # Raise a warning if the calibration circuits regularly have noisy
+            # expectation value far from the original circuit.
+            if is_far_count > len(calibration) / 2:
```
Comment (sjdilkes) [inline qermit/clifford_noise_characterisation/cdr_post.py:126]: Is 50% a reliable proportion, or should it be possible to pass this ratio as a parameter?

**Review by sjdilkes — CHANGES_REQUESTED** [review]
Graph wiring nice, just a couple of bits

File: qermit/clifford_noise_characterisation/cdr_post.py (hunk 3 above)
Comment (daniel-mills-cqc) [reply qermit/clifford_noise_characterisation/cdr_post.py:126]: Yeah, a parameter would probably be better.

File: qermit/clifford_noise_characterisation/cdr_post.py [hunk 4]
```diff
@@ -123,7 +127,7 @@ def cdr_quality_check_task(
 
             # Raise a warning if the calibration circuits regularly have noisy
             # expectation value far from the original circuit.
-            if is_far_count > len(calibration) / 2:
+            if is_far_count > len(calibration) * calibration_fraction:
                 warnings.warn(
                     "Training data regularly differers significantly from original circuit. Fit may be poor."
```
Comment (sjdilkes) [inline qermit/clifford_noise_characterisation/cdr_post.py:132]: differers -> differs

**Review by sjdilkes — DISMISSED** [review]

**Review by sjdilkes — APPROVED** [review]

**Pull request merged**